### ISSUE TITLE
False negative when using expressions in type annotations

### ISSUE THREAD

**Issue by MHDante** [microsoft/pyright#1537]
I think there's an error in the expression parser for type annotations:

The code below tries to use most available methods to define a type hint.
There are some expressions that are causing errors or producing the wrong type hint.

Or, rather, the type hint does not match the reported version by the runtime. 
When running the code, and using dataclass.fields() to get the type of the members, all of the examples below return float.

I purposely left out Annotated because it would add quite a bit of complexity.

While I would argue that all of these methods should be allowed, I understand that supporting them all might be beyond the scope of this package. However, there should perhaps be a better warning than the ones shown below and some documentation on what is/isn't allowed.

---------
## Code:

```python
from __future__ import annotations

from dataclasses import dataclass
from typing import Any
from typing_extensions import Literal,get_type_hints
typeTuple = (str, float)

type1 = typeTuple[1]


def get_type(i: Literal[0, 1]):
    return typeTuple[i]


def get_type1():
    return typeTuple[1]


class A:
    B = float

a = A

def get_nested_type():
    return A.B


a_instance = A()


def get_instance_nested_type():
    return a_instance.B


class C(A):
    pass


def get_inherited_nested_type():
    return C.B


c_instance = C()


def get_instance_inherited_nested_type():
    return c_instance.B


f = 3.2


def get_type_through_class():
    return f.__class__


class GetItemTest:
    def __getitem__(self, i: Any):
        return float


get_item_instance = GetItemTest()


def get_through_instance_get_item():
    return get_item_instance["cheem"]


class ClsGetItemTest:
    def __class_getitem__(cls, item: Any):
        return float


def get_through_class_get_item():
    return ClsGetItemTest["cheem"]


class_get_item_assmt = ClsGetItemTest["cheem"]


@dataclass
class Cheem:
    param00: float
    param01: type1
    param02: typeTuple[1]
    param03: get_type(1)
    param04: get_type1()
    param05: A.B
    param06: a.B
    param07: get_nested_type()
    param08: a_instance.B
    param09: get_instance_nested_type()
    param10: C.B
    param11: get_inherited_nested_type()
    param12: c_instance.B
    param13: get_instance_inherited_nested_type()
    param14: f.__class__
    param15: get_type_through_class()
    param16: get_item_instance["cheem"]
    param17: get_through_instance_get_item()
    param18: ClsGetItemTest["cheem"]
    param19: get_through_class_get_item()
    param20: class_get_item_assmt


cheem = Cheem(
     0.0,
     1.0,
     2.0,
     3.0,
     4.0,
     5.0,
     6.0,
     7.0,
     8.0,
     9.0,
    10.0,
    11.0,
    12.0,
    13.0,
    14.0,
    15.0,
    16.0,
    17.0,
    18.0,
    19.0,
    20.0,
)
x : float = -1.0

x = cheem.param00
x = cheem.param01
x = cheem.param02
x = cheem.param03
x = cheem.param04
x = cheem.param06
x = cheem.param05
x = cheem.param07
x = cheem.param08
x = cheem.param09
x = cheem.param10
x = cheem.param11
x = cheem.param12
x = cheem.param13
x = cheem.param14
x = cheem.param15
x = cheem.param16
x = cheem.param17
x = cheem.param18
x = cheem.param19
x = cheem.param20

hints = get_type_hints(Cheem, include_extras=True, globalns=globals(), localns=locals())

for field in hints:
    hint = hints[field]
    print(field, hint)
````
---------
## Pyright Output:

```
PS C:\Repos\pyright-test> pyright .
Loading configuration file at C:\Repos\pyright-test\pyrightconfig.json
Assuming Python platform Windows
stubPath C:\Repos\pyright-test\typings is not a valid directory.
Searching for source files
Found 11 source files
C:\Repos\pyright-test\proj1\src\typed_tuple.py
  85:14 - error: Expected class type but received "tuple[Type[str], Type[float]]" (reportGeneralTypeIssues)
  91:14 - error: Expected class type but received "A" (reportGeneralTypeIssues)
  95:14 - error: Expected class type but received "C" (reportGeneralTypeIssues)
  97:14 - error: Expected class type but received "float" (reportGeneralTypeIssues)
  99:14 - error: Expected class type but received "GetItemTest" (reportGeneralTypeIssues)
  125:5 - error: Argument of type "float" cannot be assigned to parameter "param18" of type "ClsGetItemTest" in function "__init__"
    "float" is incompatible with "ClsGetItemTest" (reportGeneralTypeIssues)
  126:5 - error: Argument of type "float" cannot be assigned to parameter "param19" of type "ClsGetItemTest" in function "__init__"
    "float" is incompatible with "ClsGetItemTest" (reportGeneralTypeIssues)
  127:5 - error: Argument of type "float" cannot be assigned to parameter "param20" of type "class_get_item_assmt" in function "__init__"
    "float" is incompatible with "ClsGetItemTest" (reportGeneralTypeIssues)
  134:5 - error: Expression of type "str | float" cannot be assigned to declared type "float"
    Type "str | float" cannot be assigned to type "float"
      "str" is incompatible with "float" (reportGeneralTypeIssues)
  149:5 - error: Expression of type "ClsGetItemTest" cannot be assigned to declared type "float"
    "ClsGetItemTest" is incompatible with "float" (reportGeneralTypeIssues)
  150:5 - error: Expression of type "ClsGetItemTest" cannot be assigned to declared type "float"
    "ClsGetItemTest" is incompatible with "float" (reportGeneralTypeIssues)
  151:5 - error: Expression of type "class_get_item_assmt" cannot be assigned to declared type "float"
    "ClsGetItemTest" is incompatible with "float" (reportGeneralTypeIssues)
12 errors, 0 warnings, 0 infos
Completed in 0.93sec
```

-------

## Python output:

```
PS C:\Repos\pyright-test> python .\proj1\src\typed_tuple.py
param00 <class 'float'>
param01 <class 'float'>
param02 <class 'float'>
param03 <class 'float'>
param04 <class 'float'>
param05 <class 'float'>
param06 <class 'float'>
param07 <class 'float'>
param08 <class 'float'>
param09 <class 'float'>
param10 <class 'float'>
param11 <class 'float'>
param12 <class 'float'>
param13 <class 'float'>
param14 <class 'float'>
param15 <class 'float'>
param16 <class 'float'>
param17 <class 'float'>
param18 <class 'float'>
param19 <class 'float'>
param20 <class 'float'>
PS C:\Repos\pyright-test>
```

**Comment by erictraut** [comment]
No, arbitrary expressions are not supported in type annotations. Refer to PEP 484 for details.

The whole idea behind a type annotation is to provide an unambiguous type so type inference is not needed. If arbitrary expressions were allowed for annotations, their types would need to be inferred.

Mypy also emits errors in these cases.

**Comment by MHDante** [comment]
Right, I'm aware of the distinction. 
I looked through documentation, (especially PEP 484) for the spec, but I found only the following:

> Annotations must be valid expressions that evaluate without raising exceptions at the time the function is defined (but see below for forward references).

> Annotations should be kept simple or static analysis tools may not be able to interpret the values. For example, dynamically computed types are unlikely to be understood. (This is an intentionally somewhat vague requirement, specific inclusions and exclusions may be added to future versions of this PEP as warranted by the discussion.)

With regards to pyright, my concern is that some expressions are supported, but others aren't. The line between the two is somewhat unclear. The expressions in this report (Up until param17) are all """Supported""" through wrapping the expression in a function. 

For example, param08 and param09 each have the same expression, yet param09 does not issue an error due to being called through a function. 

I'm trying to internalize the rules for what constitutes a valid type hint (in pyright). I could understand if only class names and special forms were allowed, but in that case, I would recommend emitting an error when a function is used. (similar to how mypy does)

For context, here is a the visual from within VSCode:

![image](https://user-images.githubusercontent.com/4652122/109132511-e8e6d480-7721-11eb-91d5-4b9285789d10.png)

**Comment by erictraut** [comment]
Ah, I thought you were saying that these expressions should be allowed (i.e. the errors are false positives). 

I agree that some of these expression forms that are apparently allowed should be flagged as errors (i.e. false negatives). Yeah, it would be reasonable for pyright to generate errors in more of these cases. It has never come up before because people don't generally attempt to use expressions in type annotations. So you're doing something very unusual here.

**Comment by erictraut** [comment]
I've added a diagnostic (error) for the case where a call expression is used within a type annotation. This will be in the next version of pyright.

Also see #1538.

**Comment by MHDante** [comment]
Hey, So I just want to share a cool thing:

I wrote this small utility that uses Annotated factories to create metadata for runtime introspection.

https://gist.github.com/MHDante/a00747f17413eead03cbbcbbcf88f6fb

It relies on a dataclass of the following format:

```
@dataclass
    class MySettings:
        listen_address: Setting(str, "host to listen on") = "0.0.0.0"
        remote_address: Setting(str, "public host address") = "localhost"
        port: Setting(int, "port to serve on") = 80

```

The function `Setting(x:Callable[[str],S])` would return an `Annotated[S,...]`. However, it would tell the type checker that it Returned a Type[T]

```

def Setting( type: Callable[[str], S], help: str = "") -> Type[S]:
    return Annotated[S, **locals()]
```

As far as the type checker was aware. This is an expression of type `T`

But the annotated info helpfuly guided by the non-existant documentation on `Setting` would be available to use.

I'm quite fond of this kind of stuff. However, the proposed change (I'm aware I proposed it 😅 ) would label these factories as an error.

Correct me if I'm wrong, but it seems like the functionality before was that if the expression in the type hint was inferred to be of type `Type[T]`, the field would be labeled to be of type `T`. 

**This does not mean that the expression is evaluated to calculate the type hint, but rather it's the inferred type of the expression, so long as the expression extends Type[T]** 
I think this is actually really cool! 

It just seems that there was some parsing logic that broke when there was a `.` or `[]` on a non-class expression. This is likely due to the special parsing (and autocomplete, maybe?) given to some special forms.

Anyway, don't let this stop you from closing the issue.  I understand that there's significant maintentance overhead on a scope expansion like this. That said... I'd like to try convincing mypy to adopt a similar implementation and I think this would lead to a lot more flexibility in typing expressions.

**Comment by erictraut** [comment]
Ah, thanks for the explanation. I found your initial bug report rather odd, but it makes much more sense now! :)

I think your idea is viable if you use `Annotated` directly (in a static form) rather than using a runtime call that returns an `Annotated` object. This might not look as nice, but it should work with existing standards and type checkers (at least once they support [PEP 593](https://www.python.org/dev/peps/pep-0593/)).

```python
@dataclass
    class MySettings:
        listen_address: Annotated[str, Setting("host to listen on")] = "0.0.0.0"
        remote_address: Annotated[str, Setting("public host address")] = "localhost"
        port: Annotated[int, Setting("port to serve on")] = 80
```

I think you'll find that you will get significant resistance from authors of the various type checkers if you propose to broaden the flavors of expressions supported within type annotations. Type annotations are meant to be statically analyzable with no ambiguity across type checkers. More complex expressions (including calls) conflict with this core requirement.

**Comment by MHDante** [comment]
Yep. That's likely What I will end up doing 😢 

My final off-topic comment:
-----
I really think pyright's implementation might be the missing piece of [PEP 484's Vague section](https://www.python.org/dev/peps/pep-0484/#acceptable-type-hints) : 

Consider this napkin-scribble spec:

In an expression of the form:

`variable_name : type_hint`

 - Infer the type (without evaluating) of the `type_hint` expression
 - The type of the expression `type_hint` must be inferrable
 - The type of the expression `type_hint` must be assignable to `Type [T]`
 - `variable_name` is assumed to be of type `T`
 - Lotta missing thinking around forward declarations... 🤔 
 
-----
I would love to try to draft an actual spec and ref implementations on my next vacation. :)

But I should post this on python/typing. Thanks for your help!

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.116, which I just published. It will also be in the next release of pylance.
